plugins/nexusoperations/frontend/handler.go

```diff
@@ -0,0 +1,309 @@
+//
+// Copyright (c) 2024 Temporal Technologies Inc.  All rights reserved.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package frontend
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"net/url"
+	"runtime/debug"
+	"strings"
+	"time"
+
+	"github.com/gorilla/mux"
+	"github.com/nexus-rpc/sdk-go/nexus"
+	"go.temporal.io/api/serviceerror"
+	"go.temporal.io/server/api/historyservice/v1"
+	"go.temporal.io/server/common/authorization"
+	"go.temporal.io/server/common/dynamicconfig"
+	"go.temporal.io/server/common/log"
+	"go.temporal.io/server/common/log/tag"
+	"go.temporal.io/server/common/metrics"
+	"go.temporal.io/server/common/namespace"
+	commonnexus "go.temporal.io/server/common/nexus"
+	"go.temporal.io/server/common/resource"
+	"go.temporal.io/server/common/rpc"
+	"go.temporal.io/server/common/rpc/interceptor"
+	"go.temporal.io/server/service/frontend/configs"
+	"go.uber.org/fx"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/status"
+)
+
+var Module = fx.Module(
+	"plugin.nexusoperations.frontend",
+	fx.Provide(ConfigProvider),
+	fx.Provide(commonnexus.NewCallbackTokenGenerator),
+	fx.Invoke(RegisterHTTPHandler),
+)
+
+var apiName = configs.CompleteNexusOperation
+
+type Config struct {
+	Enabled dynamicconfig.BoolPropertyFn
+}
+
+func ConfigProvider(coll *dynamicconfig.Collection) *Config {
+	return &Config{
+		Enabled: coll.GetBoolProperty(dynamicconfig.FrontendEnableNexusAPIs, false),
+	}
+}
+
+type HandlerOptions struct {
+	fx.In
+
+	NamespaceRegistry                    namespace.Registry
+	Logger                               log.Logger
+	MetricsHandler                       metrics.Handler
+	Config                               *Config
+	CallbackTokenGenerator               *commonnexus.CallbackTokenGenerator
+	HistoryClient                        resource.HistoryClient
+	NamespaceValidationInterceptor       *interceptor.NamespaceValidatorInterceptor
```

Comment:
I didn't read the code very carefully. I thought those interceptors are taken care of by the existing NexusHTTPHandler in frontend? I guess I was confused by all those handlers now 🤦

Reply:
Yeah, I'm also a little confused why there is a separate Nexus frontend handler just for completions.